### PR TITLE
Feat/multipass

### LINKED ISSUE #8: Must change ve_response to ve_response['inc_data']
[volume_element_response must become volume_element_response['incremental_data'] when Mike merges his branches!](https://github.com/LightForm-group/matflow-damask/pull/7#pullrequestreview-665464300)

### PR DESCRIPTION
**Pull request by Gbowker**
Add following changes to dev branch:
- Added function mapper `modify_volume_element_new_orientations` to main.py which:
1. Reads quaternion data from final increment.
2. Randomly samples `n_grains` orientations to pass to the next simulation (loading pass).

- Added function mapper `modify_volume_element_geometry` to main.py which:
1. Matrix-multiplies the VE size at the first increment by the strain rate tensor at the final increment of the simulate_loading task.
2. Passes the strain of the final increment of the previous simulate_loading task to the initial VE of the next, however without changing the grid_size. This may need to be improved.

These methods are required by task schema methods committed into [UoM-CSF-matflow branch feat/multipass task_schemas.yml](https://github.com/LightForm-group/UoM-CSF-matflow/blob/feat/multipass/task_schemas.yml) in the task `modify_volume_element`.

An example workflow has been included to make use of these additions in [UoM-CSF-matflow branch feat/multipass workflows/multipass_rolling.yml](https://github.com/LightForm-group/UoM-CSF-matflow/blob/feat/multipass/workflows/multipass_rolling.yaml). **PLEASE TEST THIS WITH DIFFERENT MATERIAL PARAMETERS BEFORE MERGING**

Thanks,

Guy

### CODE REVIEW

File: matflow_damask/main.py [hunk 1]
```diff
@@ -372,6 +371,33 @@ def modify_volume_element_add_buffer_zones(volume_element, buffer_sizes,
     }
     return out
 
+@func_mapper(task='modify_volume_element', method='new_orientations')
+def modify_volume_element_new_orientations(volume_element, volume_element_response):
+
+    n_grains = volume_element['orientations']['quaternions'].shape[0]
+    n_fragments = volume_element_response['orientations']['data']['quaternions'][-1].shape[0]
```
Comment (mikesmic) [inline matflow_damask/main.py]: Couldn't `[-1].shape[0]` just be `.shape[1]`?

**Review by mikesmic — APPROVED** [review]
Looks fine. `volume_element_response` will need to changed to `volume_element_response['incremental_data']` when other branches are merged

File: matflow_damask/main.py [hunk 2]
```diff
@@ -372,6 +371,33 @@ def modify_volume_element_add_buffer_zones(volume_element, buffer_sizes,
...
+    return out
+
+@func_mapper(task='modify_volume_element', method='geometry')
+def modify_volume_element_geometry(volume_element, volume_element_response):
+
+    print("\nold_geomsize:", volume_element['size']) # DEBUG
+    volume_element['size'] = np.matmul(volume_element_response['def_grad']['data'][-1], volume_element['size'])
+    print("\nnew_geomsize:", volume_element['size']) # DEBUG
+    
+    # return volume_element with new size ...
```
Comment (aplowman) [inline matflow_damask/main.py]: I don't think we need this comment.

File: matflow_damask/main.py [hunk 3]
```diff
@@ -372,6 +371,33 @@ def modify_volume_element_add_buffer_zones(volume_element, buffer_sizes,
...
+    n_fragments = volume_element_response['orientations']['data']['quaternions'][-1].shape[0]
+
+    old_oris = volume_element_response['orientations']['data']['quaternions'][-1]
+    random_index = np.random.randint(n_fragments, size=n_grains) ; print("randomindex array size: ", random_index.shape)
+    volume_element['orientations']['quaternions'] = old_oris[random_index, :]
+
+    print("old orientations array shape: ", old_oris.shape, "\nnew orientations array shape: ", old_oris[random_index, :].shape) # DEBUG
+
+    # return volume_element with new oris...
+    out = {'volume_element': volume_element} ; print("volume_element['orientations']['quaternions']:\n", volume_element['orientations']['quaternions'])
```
Comment (aplowman) [inline matflow_damask/main.py]: Could you either put your `print`s on a new line, or remove them?

File: matflow_damask/main.py (hunk 1 above)
Comment (Gbowker) [reply matflow_damask/main.py]: Commit: 9bc0b4b
"l:378 [-1].shape[0] now .shape[1]"

File: matflow_damask/main.py (hunk 2 above)
Comment (Gbowker) [reply matflow_damask/main.py]: Commit: 51828b4
"l:386 & l:397 removed comments"

File: matflow_damask/main.py (hunk 3 above)
Comment (Gbowker) [reply matflow_damask/main.py]: Commit: f50ebae
"l:387 removed print ve_quats"

**Comment by Gbowker** [comment]
> Looks fine. `volume_element_response` will need to changed to `volume_element_response['incremental_data']` when other branches are merged

Shall I make this an issue? Or could we just make the change in that branch now? And make a note to resolve it if it conflicts when merging?

**Comment by mikesmic** [comment]
Make an issue as a reminder I think

**Review by aplowman — APPROVED** [review]

**Comment by aplowman** [comment]
@Gbowker, feel free to add your name to the authors list in `setup.py` if you want to:

https://github.com/LightForm-group/matflow-damask/blob/c43d482b189ec99afe9670a47a980c1396c37567/setup.py#L40

**Comment by Gbowker** [comment]
Commit: 6585a09
"Added myself to authors list"

Thanks! :D

**Comment by aplowman** [comment]
> Commit: [6585a09](https://github.com/LightForm-group/matflow-damask/commit/6585a09ce99e2c33e77c9069d03a78271d79ed0e)
> "Added myself to authors list"
> 
> Thanks! :D

I like to share the blame 😂

**Pull request merged**